### DISCUSSION TITLE
Several classes not found by Android Studio 2021.2+

### DISCUSSION THREAD

**Discussion by moving-bits** [cgeo/cgeo#13606] [Q&A]
I have a persisting problem with Android Studio:

When I use either AS 2021.2 or AS 2021.3, several classes are not found by the indexer and marked red in the editor. This means I cannot jump to their source, cannot view supported parameters, and have tons of false errors.

This is an example file:
![image](https://user-images.githubusercontent.com/3754370/200117344-25f4781b-9988-41ae-8338-21ef39ea050c.png)

Some of the included libraries are affected (AndroidX, RxJava, Apache commons and VTM in this case), but not all of them.

I don't have those problems with AS 2021.1 (which is why I'm still sticking to that release, although 2021.3 has a nice new LogCat viewer).

I've tried several things already, but to no avail:
- Restarting IDE
- Wiping caches and restarting IDE
- Manually deleting cache folders generated by Gradle or IDE
- Uninstalling AS including settings folders, manually deleting/renaming remaining folders, and reinstalling

Nothing has helped so far, so I reverted again to AS 2021.1

Any ideas on how to resolve this?

**Comment by bekuno** [comment]
I had a similar problem.
I deinstalled AS, deleted (renamed) `.gradle` and `.android` in the (windows) user home.
After a fresh AS installation now AS works fine with 2021.3.1.1.

**Reply by moving-bits** [reply]
Thank for your reply. I remember you'd mentioned this some time ago. I had tried it back then, and just retried it now, but to no avail, sadly.

Somewhere AS seems to store additional info, as it remembers the last used project as well as the installed plugins, although I had uninstalled AS, deleted the folders mentioned (and more), and had restarted the computer. 🤔

**Reply by ztNFny** [reply]
C:\Users\<Your User>\AppData\Roaming\Google\.AndroidStudio4.X

**Reply by moving-bits** [reply]
Thanks for the pointer. Helped me with cleaning up a bit (worked as expected), but alas did not solve my original problem, sadly: AS started with its assistant, making me select my project (importing c:geo), but as I had deleted one folder too many, Gradle was unable to work at all. Thanks to my backup I'm back in the game. Only that the classes are still not found by AS when using 2021.3 version. :-(

**Comment by bekuno** [comment]
Maybe there is something problematic in the project directory special folders, like `.gradle`or `idea`?

**Reply by moving-bits** [reply]
I had deleted (renamed) the `.gradle` subfolder of the c:geo git project as well as some subfolders of the `idea` folder, but no change.

I have no idea what's going on here, and haven't found anything on the net about it yet :-(

**Reply by ztNFny** [reply]
sounds like a gradle issue. Can you do a gradle build from commandline? If so everything should be there but Android Studio just isn't looking into the gradle dir for the sources. Usually they should be shown in the tree in AS.

**Reply by moving-bits** [reply]
Compiling does work without problem (have tried only from within AS yet), so I guess the problem is related to AS indexing the symbols incompletely.

**Comment by fm-sys** [comment]
I had the same problem today. `Files > Invalidate Caches...` worked for me.

**Comment by moving-bits** [comment]
Still happening, and getting worse, see https://github.com/cgeo/cgeo/issues/13666#issuecomment-1327936293

**Comment by moving-bits** [comment]
I've opened an [issue with JetBrains](https://youtrack.jetbrains.com/issue/IDEA-306829/Indexing-in-Android-Studio-randomly-missing-several-classes) for this, let's see if they can find anything useful.

**Reply by Narrat** [reply]
Was the issue moved? When I try the URL it fails to load the specified issue

**Reply by moving-bits** [reply]
Issue seems to be non-public. When I follow my own link above, I get a 404 message, and only after I have logged in to YouTrack I can see the issue. But they are working on it, asking me for further info in the morning, getting answers from me in the evening :-)

**Comment by moving-bits** [comment]
For reference: I asked JetBrains which folders are involved. They directed me to the following support articles:
- for IntelliJIdea: https://intellij-support.jetbrains.com/hc/en-us/articles/206544519-Directories-used-by-the-IDE-to-store-settings-caches-plugins-and-logs
- for AndroidStudio: https://developer.android.com/studio/intro/studio-config#file_location

Will test next if I can do a clean install with that info.

**Reply by moving-bits** [reply]
Have unistalled Android Studio, removed the following folders, and reinstalled AS:
- %APPDATA%\Google\<product><version>
- %USERPROFILE%\.gradle
- %USERPROFILE%\.m2
- %USERPROFILE%\AppData\Local\Android\SDK
- %LOCALAPPDATA%\Google\<product><version>

No real progress, though, only different classes not being detected :-(

**Comment by devemux86** [comment]
Have you tried to open the project in IDEA?

**Reply by moving-bits** [reply]
(Sorry for responding so late, had some busy weeks.)

I've installed IntelliJ Community Edition 2022.3 and opened the existing c:geo project. The same classes seem to be undetected.
Did a "clean project" and rebuild then, but that did not make any difference.
The app compiles and runs fine, though.

In my "test file", only a single part of VTM is affected, the other classes are detected correctly:
![image](https://user-images.githubusercontent.com/3754370/208316910-5d170b52-4521-43ae-9626-df434ca05c8e.png)

**Comment by moving-bits** [comment]
https://github.com/cgeo/cgeo/pull/13818 might give a solution for this issue. Have not been able to test it thoroughly yet, but a first glance looks promising, at least for c:geo project.
